### DISCUSSION TITLE
Expose bump height in the Parallax node

### DISCUSSION THREAD

**Discussion by lordloki** [UPBGE/upbge#1274] [Ideas/Features for UPBGE 0.2.5+]
Currently the Parallax vector mapping node has 2 available inputs:

1) The number of steps to the parallax effect
2) The Depth of the parallax effect

![image](https://user-images.githubusercontent.com/67176876/91074509-a6293000-e60a-11ea-9b04-126f7307fb7f.png)

I am requesting that the height input be exposed to the parallax shader node, highlighted in this example image:

![image](https://user-images.githubusercontent.com/67176876/91075505-2dc36e80-e60c-11ea-8727-133109a77be4.png)

Here is a visual reference of how height offset works:
![image](https://user-images.githubusercontent.com/67176876/91081982-f8bc1980-e615-11ea-98b8-333f3b5a1e5f.png)
![image](https://user-images.githubusercontent.com/67176876/91081996-fd80cd80-e615-11ea-9efc-5bfbfb227acc.png)
![image](https://user-images.githubusercontent.com/67176876/91082004-007bbe00-e616-11ea-9e11-14d579537c2c.png)
![image](https://user-images.githubusercontent.com/67176876/91082012-0376ae80-e616-11ea-82e3-3767eb034351.png)

This is the standard behavior for height offset of a parallax shader-- options may be different in the version that is currently available.
If this is a simple code change, I would be immensely grateful for it's inclusion. I cannot express enough how useful this would be for users of the game engine as well as my personal project "The Shadows Lengthen". 

Best,
Thomas Murphy ( @Feral3d )

**Comment by Feral3d** [comment]
@lordloki  thank you for correcting my mistake.
While searching through the updates, I think I found the point where you can edit the depth of the parallax node! The "overall height" of the parallax, can be tweaked by editing "the depth to start from"

![image](https://user-images.githubusercontent.com/67176876/91346815-5fc10600-e7af-11ea-9064-05550e76974f.png)

This chunk of code can be found here: https://github.com/UPBGE/upbge/commit/686112da3bdb527447933fbf1fa0f7684ae433ca

**Comment by lordloki** [comment]
Here you have a build where you can modify that depth parameter but I can't see the same results that you showed before.
https://mega.nz/file/BwsxRLxZ#y-fXVt-0bHHsqMrZtwF3rSYYgW-Gn9KWD9q4jUPUygM

**Comment by BluePrintRandom** [comment]
[image: image.png]

**Comment by Feral3d** [comment]
@lordloki  wow this is so close!!! 

Maybe it is the 'height' of the parallax seen here: 
![image](https://user-images.githubusercontent.com/67176876/92313478-1be1b400-ef9a-11ea-88b6-a6bb56faf87d.png)

...or it's the min and max depth determined by the "depth to start from" seen here:
![image](https://user-images.githubusercontent.com/67176876/92313501-6e22d500-ef9a-11ea-99bd-f6945eec6641.png)

The lowest step of the parallax needs to be offest by the starting height value, so that rather than "cutting away" the steps, those steps move down with the depth value. 

You can see 
in this visual example how the height is moving down, but the clipping away the steps.


![Depth_Change](https://user-images.githubusercontent.com/67176876/92313594-9232e600-ef9b-11ea-9261-005db3417008.gif)

**Comment by lordloki** [comment]
A new test build: https://mega.nz/file/x1szGZCR#sgl0iEu7LGEEE6r_AeZXUvIPR5RqfsWpU9meCtTVRM4
This build behaves as your first images but with a "height" max

**Comment by Feral3d** [comment]
Ok this is about 99.9% working! 

In the most recent build I am able to bring down the depth of the of the overall parallax, but it cuts into the highest points in the parallax effect. This is the current results:
![Depth_Change_Current_Example](https://user-images.githubusercontent.com/67176876/92337673-40f92400-f079-11ea-9dd7-a0819f21c74c.gif)

Is it possible to move the parallax parallax "min" down along with the "Height" max at the same time? So this result can be achieved: 

![Depth_Change_Current_Example_please](https://user-images.githubusercontent.com/67176876/92338287-5ff9b500-f07d-11ea-8e24-6e57871dd8cc.gif)

**Comment by Feral3d** [comment]
Oh I found this!
It's entirely possible that the h in this "For" loop is cutting away the steps, as the height moves downwards:
![image](https://user-images.githubusercontent.com/67176876/92533896-7466ba80-f201-11ea-9cef-89aae7b6d15e.png)

**Comment by Feral3d** [comment]
So essentially if "height" is less than 1.0 or h, then it begins to subtract from the steps, and this is the cutting action that is seen in the test builds. A good way to test this would be to expose h as value that can be tweaked.

**Comment by lordloki** [comment]
That code was the initial code (from angelo miner) but it was modified a lot to solve several fixes and to improve other features.
Nowadays, it is very different. 
I'm trying to extract the height part from depth/stepdepth part but i wasn't successfull. I will search for other solution.

**Comment by Feral3d** [comment]
I am trying to help, but I'm having a bit of trouble building from source. If there was a way I could help you test I'm happy to try.

**Comment by Feral3d** [comment]
I have come up with a solution where you can pit the max height value against a second parallax input. In this example you can change the height of the parallax exactly as expected: 
![image](https://user-images.githubusercontent.com/67176876/92836623-a0716f80-f3aa-11ea-8728-87f84b4516c9.png)

**Comment by Feral3d** [comment]
By setting the steps to -1 and then using a white texture with a parallax node, you are able to adjust the height of the shader directly. This gives users the awesome ability to adjust the height of the parallax shader with textures and vertex painting like in this example:
![image](https://user-images.githubusercontent.com/67176876/92838463-d283d100-f3ac-11ea-946e-dbfbd7755dbe.png)

Here is the example file that you can test and edit yourself:
This hopefully helps you reverse engineer the solution!
https://www.dropbox.com/s/dasflnda5j6vvlm/TEST.blend?dl=0

**Comment by youle31** [comment]
I move this to 0.2.5 discussions
